internal/api/hooks.go

```diff
@@ -55,20 +74,219 @@ func (a *API) runHook(ctx context.Context, tx *storage.Connection, name string,
+				hookLog.Errorf("Failed to establish a connection on attempt %d with err %s", i, err)
+				time.Sleep(DefaultHTTPHookTimeout)
+				continue
+			} else if i == DefaultHTTPHookRetries-1 {
+				return nil, httpError(http.StatusGatewayTimeout, "Failed to perform webhook in time frame (%v seconds)", time.Since(startTime).Seconds())
+
+			} else {
+				return nil, internalServerError("Failed to trigger auth hook, error making HTTP request").WithInternalError(err)
+			}
+		}
+
+		switch rsp.StatusCode {
+		case http.StatusOK, http.StatusNoContent, http.StatusAccepted:
+			if rsp.Body == nil {
+				return nil, nil
+			}
+			defer rsp.Body.Close()
+			body, err := io.ReadAll(rsp.Body)
+			if err != nil {
+				return nil, err
+			}
+			if isOverSizeLimit(body) {
+				return nil, internalServerError("payload too large")
+			}
+			return body, nil
+		case http.StatusTooManyRequests, http.StatusServiceUnavailable:
+			retryAfterHeader := rsp.Header.Get("retry-after")
+			if retryAfterHeader != "true" {
+				continue
+			}
+			// TODO: maybe return errors properly for all of these from response
+			return []byte{}, tooManyRequestsError("too many requests")
+		case http.StatusBadRequest:
+			return nil, badRequestError("bad request error")
+		case http.StatusUnauthorized:
+			return []byte{}, unauthorizedError("unauthorized")
+		default:
+			return []byte{}, internalServerError("error executing hooks")
+		}
+	}
+	return nil, internalServerError("error executing hook")
+}
+
+func generateSignatures(secrets []string, msgID uuid.UUID, currentTime time.Time, inputPayload []byte) ([]string, error) {
+	// TODO(joel): Handle asymmetric case once library has been upgraded
+	var signatureList []string
+	for _, secret := range secrets {
+		if strings.HasPrefix(secret, SymmetricSignaturePrefix) {
+			trimmedSecret := strings.TrimPrefix(secret, SymmetricSignaturePrefix)
+			wh, err := standardwebhooks.NewWebhook(trimmedSecret)
+			if err != nil {
+				return nil, err
+			}
+			signature, err := wh.Sign(msgID.String(), currentTime, inputPayload)
+			if err != nil {
+				return nil, err
+			}
+			signatureList = append(signatureList, signature)
+		} else {
+			return nil, errors.New("invalid signature format")
+		}
+	}
+	return signatureList, nil
+}
+
+func watchForConnection(req *http.Request) (*connectionWatcher, *http.Request) {
+	w := new(connectionWatcher)
+	t := &httptrace.ClientTrace{
+		GotConn: w.GotConn,
+	}
+
+	req = req.WithContext(httptrace.WithClientTrace(req.Context(), t))
+	return w, req
+}
+
+type connectionWatcher struct {
+	gotConn bool
+}
+
+func (c *connectionWatcher) GotConn(_ httptrace.GotConnInfo) {
+	c.gotConn = true
+}
+
+func isOverSizeLimit(payload []byte) bool {
+	// As per spec, payloads should be under 20KB https://github.com/standard-webhooks/standard-webhooks/blob/main/spec/standard-webhooks.md#payload-size
+	const maxSizeKB = 20 * 1024
+	return len(payload) > maxSizeKB
+}
+
+func validateHTTPHook(uri string) error {
```

Comment:
Please move to config validation. Only allow HTTPS except localhost which can be HTTP.